src/core/tools/common/logger_proxy.h

```diff
@@ -0,0 +1,73 @@
+/*
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2015 Microsoft Corporation
+ *
+ * -=- Robust Distributed System Nucleus (rDSN) -=-
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+/*
+ * Description:
+ *   logger_proxy is a proxy of logger_provider. The reference to the target
+ * object of logger_provider is controlled by logger_proxy.
+ */
+
+#pragma once
+
+#include <dsn/tool-api/logging_provider.h>
+
+namespace dsn {
+namespace tools {
+
+class logger_proxy : public logging_provider
```

Comment:
I think let the logger depends on config reader is a best practise, because log reader needs to be initialized by dsn runtime.
The logger_proxy is a interface of all of child logger type. It is used to control and manage access to an object of logger child type, so these logger child type don't have to be exposed outside.

Reply:
config reader belongs to dsn_utils, logger belongs to dsn_utils too.
So why do you think that logger depends on config reader is not reasonable?

Reply:
Here is my thought: 
If the config file is not load by dsn runtime, we can get nothing from it. In other words, if the rdsn runtime is not running, the config reader doesn`t work. So If we let logger depends on the config reader, it depends on the dsn runtime indirectly.